stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,780 @@
+use std::collections::BTreeMap;
+
+use crate::crypto::{Coordinatable, Error as CryptoError, OperationResult};
+use frost_signer::{
+    net::Message,
+    signing_round::{DkgBegin, DkgPublicShare, MessageTypes, Signable, NonceRequest, NonceResponse, SignatureShareRequest},
+};
+use hashbrown::HashSet;
+use slog::{slog_info, slog_warn};
+use stacks_common::{error, info, warn};
+use wsts::{
+    Point, Scalar,
+    errors::AggregatorError,
+    taproot::{
+        Error as TaprootError, SchnorrProof,
+    },
+    common::{
+        PolyCommitment, PublicNonce, Signature, SignatureShare,
+    },
+    v1,
+};
+
+#[derive(thiserror::Error, Debug)]
+/// The error type for the coordinator
+pub enum Error {
+    /// A bad state change was made
+    #[error("Bad State Change: {0}")]
+    BadStateChange(String),
+    /// SignatureAggregator error
+    #[error("Aggregator: {0}")]
+    Aggregator(AggregatorError),
+    /// Taproot error
+    #[error("Taproot")]
+    Taproot(TaprootError),
+    /// Schnorr proof failed to verify
+    #[error("Schnorr Proof failed to verify")]
+    SchnorrProofFailed,
+}
+
+impl From<AggregatorError> for Error {
+    fn from(err: AggregatorError) -> Self {
+        Error::Aggregator(err)
+    }
+}
+
+impl From<TaprootError> for Error {
+    fn from(err: TaprootError) -> Self {
+        Error::Taproot(err)
+    }
+}
+
+/// The coordinator for the FROST algorithm
+pub struct Coordinator {
+    current_dkg_id: u64,
+    current_dkg_public_id: u64,
+    current_sign_id: u64,
+    current_sign_nonce_id: u64,
+    total_signers: u32, // Assuming the signers cover all id:s in {1, 2, ..., total_signers}
+    total_keys: u32,
+    threshold: u32,
+    dkg_public_shares: BTreeMap<u32, DkgPublicShare>,
+    public_nonces: BTreeMap<u32, NonceResponse>,
+    signature_shares: BTreeMap<u32, Vec<SignatureShare>>,
+    aggregate_public_key: Point,
+    signature: Signature,
+    schnorr_proof: SchnorrProof,
+    message_private_key: Scalar,
+    ids_to_await: HashSet<u32>,
+    message: Vec<u8>,
+    state: State,
+}
+
+impl Coordinator {
+    /// Create a new coordinator
+    pub fn new(total_signers: u32, total_keys: u32, threshold: u32, message_private_key: Scalar) -> Self {
+        Self {
+            current_dkg_id: 0,
+            current_dkg_public_id: 0,
+            current_sign_id: 0,
+            current_sign_nonce_id: 0,
+            total_signers,
+            total_keys,
+            threshold,
+            dkg_public_shares: Default::default(),
```

Comment:
It seems these fields (`dkg_public_shares`, `public_nonces`, `signature_shares`, `aggregate_public_key`, `signature`, and `schnorr_proof`) are generated as part of either DKG or a signing round.  If so, then would it make sense to treat them as `Option<..>`, so it's unambiguous as to whether or not they have been calculated as part of this DKG or signing round?

Reply:
I think that's a very good idea.  Let me know if you want me to do it as part of this PR or if it can wait for the follow on.

Reply:
Either way is fine.  If you do it in a subsequent PR, then can you create an issue for it?  Thanks!

Reply:
Issue made :D https://github.com/stacks-network/stacks-blockchain/issues/3925